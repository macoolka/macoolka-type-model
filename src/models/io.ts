/**
 * Code generated by github.com/macoolka/macoolka-gen-model, DO NOT EDIT.
 */
import * as t from 'macoolka-io'
export const MBasicScalar = t.keyof({"string":"","number":"","boolean":"","int":"","datetime":"","json":""})
export const MStringFormat = t.keyof({"UUID":"","EMail":"","IPV4":"","IPV6":"","URL":""})
export const MIdentifierKind = t.keyof({"field":"","typealias":"","interface":"","class":"","export":"","function":"","param":"","constant":"","module":"","method":"","staticmethod":""})
export const MNameable =t.type({
  name:t.stringMaxLength(64),
})
 
export const MIdentifier =t.type({
  name:MIdentifierKind,
})
 
export const MDocumentable =t.intersection([  t.type({
    /**
     * That this is no longer the preferred way.
     */
    deprecated:t.withDefault(t.boolean,false),
    description:t.withDefault(t.array(t.string),[]),
    ignore:t.withDefault(t.boolean,false),
    examples:t.withDefault(t.array(t.string),[]),
    /**
     * That reason of deprecation.
     */
    reason:t.withDefault(t.array(t.string),[]),
    path:t.withDefault(t.array(t.string),[]),
  })
,  t.partial({
    descriptions:t.record(t.string,t.any),
    /**
     * When was this feature added.
     */
    since:t.string,
    signature:t.string,
  })
])

export const MExportable =t.type({
  /**
   * That this is no longer the preferred way.
   */
  isExported:t.withDefault(t.boolean,true),
})
 
export const MScalable =t.type({
  /**
   * The Node can be marked as undefined. 
   */
  maybe:t.withDefault(t.boolean,false),
  /**
   * The field has the many multiplicity will also be marked.
   */
  isArray:t.withDefault(t.boolean,false),
  /**
   * The means that no item in the list can be null
   */
  isArrayRequired:t.withDefault(t.boolean,false),
  /**
   * The means that defaultValue is []
   */
  defaultEmptyArray:t.withDefault(t.boolean,true),
})
 
/**
 * Like a Boolean an Enum can have one of a predefined set of values.
 * The difference is that you can define the possible values
 * (whereas for a Boolean the options are restriced to true and false).
 * For example you could specify how an article should be formatted
 * by creating an Enum with the possible values COMPACT, WIDE and COVER.
 */
export const MEnumScalar =t.intersection([MScalable,t.intersection([  t.type({
    /**
     * The _kind is a tag that check which Sclar be used.
     */
    _kind:t.literal('enum'),
    values:t.withDefault(t.array(t.string),[]),
  })
,  t.partial({
    /**
     * You can set a default value for non-list scalar fields.
     * The value will be applied to newly created records when no value was supplied during the create-operation.
     */
    defaultValue:t.string,
  })
])
],'MEnumScalar')

export const MJsonScalar =t.intersection([MScalable,t.intersection([  t.type({
    /**
     * The _kind is a tag that check which Sclar be used.
     */
    _kind:t.literal('json'),
  })
,  t.partial({
    /**
     * You can set a default value for non-list scalar fields.
     * The value will be applied to newly created records when no value was supplied during the create-operation.
     */
    defaultValue:t.record(t.string,t.any),
  })
])
],'MJsonScalar')

/**
 * A String holds text. This is the type you would use for a username,
 * the content of a blog post or anything else that is best represented as text.
 */
export const MStringScalar =t.intersection([MScalable,t.intersection([  t.type({
    _kind:t.literal('string'),
  })
,  t.partial({
    /**
     * You can set a default value for non-list scalar fields.
     * The value will be applied to newly created records when no value was supplied during the create-operation.
     */
    defaultValue:t.string,
    minLength:t.int,
    maxLength:t.int,
    pattern:t.string,
    format:MStringFormat,
  })
])
],'MStringScalar')

/**
 * A Number is a number that can have decimals.
 * Use this to store values such as the price of an item in a store or the result of complex calculations.
 */
export const MNumberScalar =t.intersection([MScalable,t.intersection([  t.type({
    /**
     * The _kind is a tag that check which Sclar be used.
     */
    _kind:t.literal('number'),
  })
,  t.partial({
    /**
     * You can set a default value for non-list scalar fields.
     * The value will be applied to newly created records when no value was supplied during the create-operation.
     */
    defaultValue:t.number,
    maximum:t.number,
    minimum:t.number,
  })
])
],'MNumberScalar')

/**
 * An Int is a number that cannot have decimals. 
 * Use this to store values such as the weight of an ingredient required for a recipe or the minimum age for an event.
 */
export const MIntScalar =t.intersection([MScalable,t.intersection([  t.type({
    /**
     * The _kind is a tag that check which Sclar be used.
     */
    _kind:t.literal('int'),
  })
,  t.partial({
    /**
     * You can set a default value for non-list scalar fields.
     * The value will be applied to newly created records when no value was supplied during the create-operation.
     */
    defaultValue:t.number,
    maximum:t.int,
    minimum:t.int,
  })
])
],'MIntScalar')

/**
 * A Boolean can have the value true or false.
 * This is useful to keep track of settings such as whether the user wants to receive an email newsletter
 * or if a recipe is appropriate for vegetarians.
 */
export const MBooleanScalar =t.intersection([MScalable,t.intersection([  t.type({
    /**
     * The _kind is a tag that check which Sclar be used.
     */
    _kind:t.literal('boolean'),
  })
,  t.partial({
    /**
     * You can set a default value for non-list scalar fields.
     * The value will be applied to newly created records when no value was supplied during the create-operation.
     */
    defaultValue:t.boolean,
  })
])
],'MBooleanScalar')

/**
 * The DateTime type can be used to store date and/or time values.
 * A good example might be a person's date of birth or the time/data when a specific event is happening.
 */
export const MDateTimeScalar =t.intersection([MScalable,t.intersection([  t.type({
    /**
     * The _kind is a tag that check which Sclar be used.
     */
    _kind:t.literal('datetime'),
  })
,  t.partial({
    /**
     * You can set a default value for non-list scalar fields.
     * The value will be applied to newly created records when no value was supplied during the create-operation.
     * Now mean current time.
     */
    defaultValue:t.literal('now'),
  })
])
],'MDateTimeScalar')

/**
 * The _kind is a tag that check which Sclar be used.
 */
export const MKindScalar =t.intersection([MScalable,t.intersection([  t.type({
    /**
     * The _kind is a tag that check which Sclar be used.
     */
    _kind:t.literal('kind'),
    value:t.string,
  })
,  t.partial({
    defaultValue:t.string,
  })
])
],'MKindScalar')

/**
 * A type give a value of model types.
 */
export const MTypeScalar =t.intersection([MScalable,t.intersection([  t.type({
    _kind:t.literal('type'),
    /**
     * Model type name.
     */
    value:t.string,
  })
,  t.partial({
    defaultValue:t.string,
  })
])
],'MTypeScalar')

/**
 * A union type describes a value that can be one of several types.
 */
export const MTypeUnionScalar =t.intersection([MScalable,t.intersection([  t.type({
    _kind:t.literal('typeUnion'),
    /**
     * Model type name.
     */
    values:t.withDefault(t.array(t.string),[]),
  })
,  t.partial({
    defaultValue:t.string,
  })
])
],'MTypeUnionScalar')

/**
 * A intersection type describes a value that can be one of several types.
 */
export const MTypeIntersectionScalar =t.intersection([MScalable,t.intersection([  t.type({
    _kind:t.literal('typeIntersection'),
    /**
     * Model type name.
     */
    values:t.withDefault(t.array(t.string),[]),
  })
,  t.partial({
    defaultValue:t.string,
  })
])
],'MTypeIntersectionScalar')

export const MValueable =t.type({
  /**
   *  The type of this Node. A value of type indicates a scalar type.
   */
  type:t.withDefault(t.union([MEnumScalar,MStringScalar,MNumberScalar,MIntScalar,MBooleanScalar,MJsonScalar,MDateTimeScalar,MKindScalar,MTypeScalar,MTypeUnionScalar,MTypeIntersectionScalar,MBasicScalar])
,"string"),
  /**
   * The Node can be marked as required (also referred to as "non-nullable"). 
   */
  required:t.withDefault(t.boolean,false),
})
 
/**
 *  Represents a field in the datamodel.
 * ModelField is referenced by its name and is either scalar or a relation field.
 */
export const MVariable =t.intersection([MDocumentable,MValueable,MNameable,t.type({
})
 ],'MVariable')

/**
 *  Represents a field in the datamodel.
 * ModelField is referenced by its name and is either scalar or a relation field.
 */
export const MField =t.intersection([MVariable,t.type({
  _kind:t.withDefault(t.literal('field'),"field"),
  /**
   * Setting the unique constraint makes sure that two records of the model in question cannot have the same value for a certain field. 
   */
  unique:t.withDefault(t.boolean,false),
  /**
   * Id only one in a model"). 
   */
  id:t.withDefault(t.boolean,false),
  readonly:t.withDefault(t.boolean,false),
  order:t.withDefault(t.boolean,true),
  /**
   * That field exclusive when create operator.
   */
  exclusiveCreate:t.withDefault(t.boolean,false),
  /**
   * That field exclusive when create update.
   */
  exclusiveUpdate:t.withDefault(t.boolean,false),
  /**
   * That field exclusive in where condition.
   */
  exclusiveWhere:t.withDefault(t.boolean,false),
  /**
   * That field exclusive in search result.
   */
  exclusiveSearch:t.withDefault(t.boolean,false),
  /**
   * That field exclusive in load result.
   */
  exclusiveLoad:t.withDefault(t.boolean,false),
})
 ],'MField')

/**
 * ModelParam is referenced by its name and is either scalar or a relation field.
 */
export const MParam =t.intersection([MVariable,t.type({
  _kind:t.withDefault(t.literal('param'),"param"),
})
 ],'MParam')

/**
 * Method is referenced by its name and is either scalar or a relation field.
 */
export const MFunctional =t.intersection([MDocumentable,MValueable,MNameable,t.type({
  /**
   * Field can be marked as required (also referred to as "non-nullable"). 
   */
  params:t.withDefault(t.array(MParam),[]),
  /**
   * The means that the method return void 
   */
  returnVoid:t.withDefault(t.boolean,false),
})
 ],'MFunctional')

/**
 * Method is referenced by its name and is either scalar or a relation field.
 */
export const MMethod =t.intersection([MFunctional,t.type({
  _kind:t.withDefault(t.literal('method'),"method"),
})
 ],'MMethod')

/**
 * Method is referenced by its name and is either scalar or a relation field.
 */
export const MStaticMethod =t.intersection([MFunctional,t.type({
  _kind:t.withDefault(t.literal('staticmethod'),"staticmethod"),
})
 ],'MStaticMethod')

/**
 * Method is referenced by its name and is either scalar or a relation field.
 */
export const MFunction =t.intersection([MFunctional,MExportable,t.type({
  _kind:t.withDefault(t.literal('function'),"function"),
})
 ],'MFunction')

/**
 * Method is referenced by its name and is either scalar or a relation field.
 */
export const MConstant =t.intersection([MVariable,MExportable,t.type({
  _kind:t.withDefault(t.literal('constant'),"constant"),
})
 ],'MConstant')

/**
 * Method is referenced by its name and is either scalar or a relation field.
 */
export const MExport =t.intersection([MDocumentable,MNameable,t.type({
  _kind:t.withDefault(t.literal('export'),"export"),
})
 ],'MExport')

/**
 * ModelType consist of multiple fields and typically represent entities from your application domain (e.g. User, Car, Order). 
 * An Model Type defines the structure for one model in your datamodel.
 * It is used to represent entities from your application domain.
 */
export const MTypeable =t.intersection([MDocumentable,MExportable,MNameable,t.type({
  fields:t.withDefault(t.array(MField),[]),
  methods:t.withDefault(t.array(MMethod),[]),
})
 ],'MTypeable')

/**
 * ModelType consist of multiple fields and typically represent entities from your application domain (e.g. User, Car, Order). 
 * An Model Type defines the structure for one model in your datamodel.
 * It is used to represent entities from your application domain.
 */
export const MInterface =t.intersection([MTypeable,t.type({
  _kind:t.withDefault(t.literal('interface'),"interface"),
  implements:t.withDefault(t.array(t.string),[]),
})
 ],'MInterface')

/**
 * ModelType consist of multiple fields and typically represent entities from your application domain (e.g. User, Car, Order). 
 * An Model Type defines the structure for one model in your datamodel.
 * It is used to represent entities from your application domain.
 */
export const MClass =t.intersection([MTypeable,t.type({
  _kind:t.withDefault(t.literal('class'),"class"),
  implements:t.withDefault(t.array(t.string),[]),
  staticMethods:t.withDefault(t.array(MStaticMethod),[]),
})
 ],'MClass')

/**
 * ModelType consist of multiple fields and typically represent entities from your application domain (e.g. User, Car, Order). 
 * An Model Type defines the structure for one model in your datamodel.
 * It is used to represent entities from your application domain.
 */
export const MTypeAlias =t.intersection([MTypeable,MValueable,t.type({
  _kind:t.withDefault(t.literal('typealias'),"typealias"),
})
 ],'MTypeAlias')

/**
 * Define the underlying data model schema.
 */
export const MModule =t.intersection([MDocumentable,MNameable,t.type({
  file:t.withDefault(t.boolean,false),
  _kind:t.withDefault(t.literal('module'),"module"),
  interfaces:t.withDefault(t.array(MInterface),[]),
  typealiases:t.withDefault(t.array(MTypeAlias),[]),
  classes:t.withDefault(t.array(MClass),[]),
  functions:t.withDefault(t.array(MFunction),[]),
  exports:t.withDefault(t.array(MExport),[]),
  constants:t.withDefault(t.array(MConstant),[]),
  idUnique:t.withDefault(t.boolean,false),
})
 ],'MModule')



