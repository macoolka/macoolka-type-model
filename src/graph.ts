
/**
 * @file
 */
import { Graph, GraphVertex, GraphEdge } from 'macoolka-algorithms/lib/graph'

import { MModule, MField, MInterface, MTypeAlias } from './models/Module'
import { isTypeScalar, isTypeUnionScalar } from './predicate'
import * as  A from 'fp-ts/lib/Array'
import { pipe } from 'fp-ts/lib/pipeable'
import { foreach } from './helper'
import * as O from 'fp-ts/lib/Option'
import * as Ord from 'fp-ts/lib/Ord'
import { Message } from './i18n'
import * as E from 'fp-ts/lib/Either'
import _topologicalSort from 'macoolka-algorithms/lib/topologicalSort'
/**
 * a simple graph 
 * @desczh
 * 一种简单的图结构
 */
export interface Vis {
    nodes: {
        id: string,
        label: String,
    }[],
    edges: {
        from: string,
        to: string,
    }[]
}
/**
 * Build Vis from Graph
 * @desczh
 * 用Graph建立Vis
 * @since 0.2.0
 * 
 */
export const createVis = (a: Graph): Vis => ({
    nodes: pipe(
        a.getAllVertices(),
        A.map(a => ({ id: a.getKey(), label: a.toString() }))
    ),
    edges: pipe(
        a.getAllEdges(),
        A.map(a => ({
            from: a.startVertex.getKey(),
            to: a.endVertex.getKey()
        }))
    )

})
/**
 * Build Graph with MModule
 * @desczh
 * 用MModule建立Graph
 * @since 0.2.0
 */
export const buildGraph = (as: MModule): E.Either<Message[], Graph> => {
    const graphVertexs: GraphVertex<string>[] = []
    const graphEdges: GraphEdge<string>[] = []
    const errors: Message[] = []

    const addVertex = (name: string) => graphVertexs.map(a => a.getKey()).includes(name)
        ? errors.push({ id: 'macoolka.data-model.typeNameRepeat', value: { model: name, name: '' } })
        : graphVertexs.push(new GraphVertex(name))

    const addGraphEdge = (from: string, to: string) => {
        const fromOption = pipe(
            graphVertexs,
            A.findFirst(a => a.value === from),

        )
        const toOption = pipe(
            graphVertexs,
            A.findFirst(a => a.value === to),
        )
        pipe(
            fromOption,
            O.map(a => {
                pipe(
                    toOption,
                    O.map(b => {
                        const edge = new GraphEdge(a, b)
                        pipe(
                            graphEdges.every(a => a.getKey() !== edge.getKey()),

                            //    A.every(a => a.getKey() !== edge.getKey()),
                            value => value ? graphEdges.push(edge) : void 0

                        )

                    }),
                    O.getOrElse(() => {
                        errors.push({ id: 'macoolka.data-model.typeNotFound', value: { model: from, name: '' } })
                    })
                )
            }),

            O.getOrElse(() => {
                errors.push({ id: 'macoolka.data-model.typeNotFound', value: { model: from, name: '' } })
            })

        )
    }


    foreach(as, {

        model: (model: MInterface, schema: MModule) => {
            addVertex(model.name)
        },
        typealiases: (a: MTypeAlias, schema: MModule) => {
            addVertex(a.name)

        },

    })
    foreach(as, {

        type: (type: MField['type'], field: MField, model: MInterface, schema: MModule) => {
            pipe(
                type,
                O.fromPredicate(isTypeScalar),
                O.map(a =>

                    addGraphEdge(model.name, a.value)

                )
            )
            pipe(
                type,
                O.fromPredicate(isTypeUnionScalar),
                O.map(a =>
                    pipe(
                        a.values,
                        A.map(type => addGraphEdge(model.name, type))

                    )
                )
            )
        },

        implement: (name: string, model: MInterface, schema: MModule) => {
            addGraphEdge(model.name, name)
        },
    })
    return pipe(
        errors,
        E.fromPredicate(a => a.length === 0, () => errors),
        E.map(() => {
            const graph = new Graph(true);
            graphEdges.map(a => graph.addEdge(a))
            graphVertexs.map(a => graph.addVertex(a))
            return graph
        })
    )

}
export  type IOType = MInterface | MTypeAlias
export const topologicalSort = (a: MModule) => {
   
    let orders: Array<IOType> = [];

    orders = orders.concat(a.interfaces)
    orders = orders.concat(a.typealiases)

    const ordMoudle = (names: string[]): Ord.Ord<IOType> => ({
        equals: (a, b) => a.name === b.name,
        compare: (a, b) => {
            const aIndex = pipe(
                names,
                A.findIndex(name => name === a.name),
                O.getOrElse(() => -1)
            )
            const bIndex = pipe(
                names,
                A.findIndex(name => name === b.name),
                O.getOrElse(() => -1)
            )
            return Ord.ordNumber.compare(bIndex,aIndex)
        }
    })
    const valid = (names: string[]): E.Either<Message[], string[]> => pipe(
        orders,
        A.map(b => {
            return pipe(
                names,
                A.findIndex(name => name === b.name),
                E.fromOption(() => ({ id: 'macoolka.data-model.typeNotFound', value: { value: { model: b.name, name: '' } } as Message }))
            )
        }),
        A.lefts,
        a => a.length > 0 ? E.left(a) : E.right(names),
        b => b as E.Either<Message[], string[]>
    )
    //   orders=a.interfaces
    return pipe(
        a,
        buildGraph,
        E.map(_topologicalSort),
        E.chain(a =>
            pipe(
                a.map(value => value.value),
                valid,
                E.map(names => pipe(
                    names,
                    ordMoudle,
                    A.sort,
                  
                    b => b(orders)
                )),

            )
        )
    )
}
