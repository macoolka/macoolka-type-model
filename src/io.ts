import { MModule as InputModule } from './models/io'
import * as t from 'macoolka-io'
import { validateFieldName, validateModelName, ValidationResult, validateId } from './generate/validation'
import { array } from 'macoolka-collection'
import { pipe } from 'fp-ts/lib/pipeable'
import { foreach } from './helper'
export * from 'macoolka-io'
import { MModule as IModule } from './models/Module'
export * from './models/io'
import { I18NOption } from 'macoolka-i18n'
import { defaultOption } from './i18n'
import { merge } from 'macoolka-functions'
import * as E from 'fp-ts/lib/Either'

export const Module = new t.Type<IModule, IModule, unknown>(
    'Module',
    (a): a is IModule => InputModule.is(a),
    (b: any, c) => pipe(
        InputModule.validate(b, c),
        E.chain((u: any) => {
            const result: ValidationResult<any>[] = []

            foreach(u, {
                model: (model) => {
                    if (u.idUnique === true) {
                        result.push(validateId(model))
                    }
                    result.push(validateFieldName(model))
                },
                schema: (schema) => {
                    if (1 != 1)
                        result.push(validateModelName(schema))
                }
            })

            return pipe(
                result,
                array.lefts,
                array.flatten,
                as => as.length === 0
                    ? t.success(u)
                    : t.failMessage(as, c)
            )
        })
    )
    ,
    a => a
)

export const show = (errors: t.Errors) => (o?: I18NOption) => {
    return t.show(errors)({i18n:merge(defaultOption, o)})
}
