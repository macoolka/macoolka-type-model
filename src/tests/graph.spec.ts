import { buildGraph, createVis, } from '../graph'
import Basic from './fixtures/Basic'
import schema from '../schema'
import * as E from 'fp-ts/lib/Either'
import { pipe } from 'fp-ts/lib/pipeable'
import { writeFileSync } from 'fs'
import { Module } from '../io'
import _topologicalSort from 'macoolka-algorithms/lib/topologicalSort'
describe('graph', () => {
    it('build Schema Graph', () => {
        pipe(
            schema,
            Module([]).decode,
            as => {
                expect(E.isRight(as)).toBeTruthy();
                return as;
            },
            E.map(a =>
                pipe(
                    buildGraph([])(a),
                    as => {
                        expect(E.isRight(as)).toBeTruthy();
                        
                        return as;
                    },
                    E.map(a => {
                        const result=_topologicalSort(a).map(value=>value.value);
                    
                        writeFileSync(__dirname + '/schema.topologicalSort.json', JSON.stringify(result))
                        writeFileSync(__dirname + '/schema.vis.json', JSON.stringify(createVis(a)))
                       
                        expect(createVis(a)).toMatchSnapshot()
                    })
                ),
            )
        )
    })
   
    it('buildGraph', () => {
        pipe(
            Basic,
            Module([]).decode,
            as => {
                expect(E.isRight(as)).toBeTruthy();
                return as;
            },
            E.map(a =>
                pipe(
                    buildGraph([])(a),
                    as => {
                        expect(E.isRight(as)).toBeTruthy();
                        return as;
                    },
                    E.map(a => {
                        writeFileSync(__dirname + '/vis.json', JSON.stringify(createVis(a)))
                        expect(createVis(a)).toMatchSnapshot()
                    })
                ),
            )
        )
    })
})