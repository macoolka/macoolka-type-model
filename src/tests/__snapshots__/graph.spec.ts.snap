// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`graph build Schema Graph 1`] = `
Object {
  "edges": Array [
    Object {
      "from": "MEnumScalar",
      "to": "MScalable",
    },
    Object {
      "from": "MJsonScalar",
      "to": "MScalable",
    },
    Object {
      "from": "MStringScalar",
      "to": "MStringFormat",
    },
    Object {
      "from": "MStringScalar",
      "to": "MScalable",
    },
    Object {
      "from": "MNumberScalar",
      "to": "MScalable",
    },
    Object {
      "from": "MIntScalar",
      "to": "MScalable",
    },
    Object {
      "from": "MBooleanScalar",
      "to": "MScalable",
    },
    Object {
      "from": "MDateTimeScalar",
      "to": "MScalable",
    },
    Object {
      "from": "MKindScalar",
      "to": "MScalable",
    },
    Object {
      "from": "MTypeScalar",
      "to": "MScalable",
    },
    Object {
      "from": "MTypeUnionScalar",
      "to": "MScalable",
    },
    Object {
      "from": "MTypeIntersectionScalar",
      "to": "MScalable",
    },
    Object {
      "from": "MValueable",
      "to": "MScalars",
    },
    Object {
      "from": "MValueable",
      "to": "MBasicScalar",
    },
    Object {
      "from": "MVariable",
      "to": "MDocumentable",
    },
    Object {
      "from": "MVariable",
      "to": "MValueable",
    },
    Object {
      "from": "MVariable",
      "to": "MNameable",
    },
    Object {
      "from": "MField",
      "to": "MVariable",
    },
    Object {
      "from": "MParam",
      "to": "MVariable",
    },
    Object {
      "from": "MFunctional",
      "to": "MParam",
    },
    Object {
      "from": "MFunctional",
      "to": "MDocumentable",
    },
    Object {
      "from": "MFunctional",
      "to": "MValueable",
    },
    Object {
      "from": "MFunctional",
      "to": "MNameable",
    },
    Object {
      "from": "MMethod",
      "to": "MFunctional",
    },
    Object {
      "from": "MStaticMethod",
      "to": "MFunctional",
    },
    Object {
      "from": "MFunction",
      "to": "MFunctional",
    },
    Object {
      "from": "MFunction",
      "to": "MExportable",
    },
    Object {
      "from": "MConstant",
      "to": "MVariable",
    },
    Object {
      "from": "MConstant",
      "to": "MExportable",
    },
    Object {
      "from": "MExport",
      "to": "MDocumentable",
    },
    Object {
      "from": "MExport",
      "to": "MNameable",
    },
    Object {
      "from": "MTypeable",
      "to": "MField",
    },
    Object {
      "from": "MTypeable",
      "to": "MMethod",
    },
    Object {
      "from": "MTypeable",
      "to": "MDocumentable",
    },
    Object {
      "from": "MTypeable",
      "to": "MExportable",
    },
    Object {
      "from": "MTypeable",
      "to": "MNameable",
    },
    Object {
      "from": "MInterface",
      "to": "MTypeable",
    },
    Object {
      "from": "MClass",
      "to": "MStaticMethod",
    },
    Object {
      "from": "MClass",
      "to": "MTypeable",
    },
    Object {
      "from": "MTypeAlias",
      "to": "MTypeable",
    },
    Object {
      "from": "MTypeAlias",
      "to": "MValueable",
    },
    Object {
      "from": "MModule",
      "to": "MInterface",
    },
    Object {
      "from": "MModule",
      "to": "MTypeAlias",
    },
    Object {
      "from": "MModule",
      "to": "MClass",
    },
    Object {
      "from": "MModule",
      "to": "MFunction",
    },
    Object {
      "from": "MModule",
      "to": "MExport",
    },
    Object {
      "from": "MModule",
      "to": "MConstant",
    },
    Object {
      "from": "MModule",
      "to": "MDocumentable",
    },
    Object {
      "from": "MModule",
      "to": "MNameable",
    },
  ],
  "nodes": Array [
    Object {
      "id": "MEnumScalar",
      "label": "MEnumScalar",
    },
    Object {
      "id": "MScalable",
      "label": "MScalable",
    },
    Object {
      "id": "MJsonScalar",
      "label": "MJsonScalar",
    },
    Object {
      "id": "MStringScalar",
      "label": "MStringScalar",
    },
    Object {
      "id": "MStringFormat",
      "label": "MStringFormat",
    },
    Object {
      "id": "MNumberScalar",
      "label": "MNumberScalar",
    },
    Object {
      "id": "MIntScalar",
      "label": "MIntScalar",
    },
    Object {
      "id": "MBooleanScalar",
      "label": "MBooleanScalar",
    },
    Object {
      "id": "MDateTimeScalar",
      "label": "MDateTimeScalar",
    },
    Object {
      "id": "MKindScalar",
      "label": "MKindScalar",
    },
    Object {
      "id": "MTypeScalar",
      "label": "MTypeScalar",
    },
    Object {
      "id": "MTypeUnionScalar",
      "label": "MTypeUnionScalar",
    },
    Object {
      "id": "MTypeIntersectionScalar",
      "label": "MTypeIntersectionScalar",
    },
    Object {
      "id": "MValueable",
      "label": "MValueable",
    },
    Object {
      "id": "MScalars",
      "label": "MScalars",
    },
    Object {
      "id": "MBasicScalar",
      "label": "MBasicScalar",
    },
    Object {
      "id": "MVariable",
      "label": "MVariable",
    },
    Object {
      "id": "MDocumentable",
      "label": "MDocumentable",
    },
    Object {
      "id": "MNameable",
      "label": "MNameable",
    },
    Object {
      "id": "MField",
      "label": "MField",
    },
    Object {
      "id": "MParam",
      "label": "MParam",
    },
    Object {
      "id": "MFunctional",
      "label": "MFunctional",
    },
    Object {
      "id": "MMethod",
      "label": "MMethod",
    },
    Object {
      "id": "MStaticMethod",
      "label": "MStaticMethod",
    },
    Object {
      "id": "MFunction",
      "label": "MFunction",
    },
    Object {
      "id": "MExportable",
      "label": "MExportable",
    },
    Object {
      "id": "MConstant",
      "label": "MConstant",
    },
    Object {
      "id": "MExport",
      "label": "MExport",
    },
    Object {
      "id": "MTypeable",
      "label": "MTypeable",
    },
    Object {
      "id": "MInterface",
      "label": "MInterface",
    },
    Object {
      "id": "MClass",
      "label": "MClass",
    },
    Object {
      "id": "MTypeAlias",
      "label": "MTypeAlias",
    },
    Object {
      "id": "MModule",
      "label": "MModule",
    },
    Object {
      "id": "MIdentifierKind",
      "label": "MIdentifierKind",
    },
  ],
}
`;

exports[`graph buildGraph 1`] = `
Object {
  "edges": Array [
    Object {
      "from": "O1",
      "to": "E1",
    },
    Object {
      "from": "O1",
      "to": "E2",
    },
    Object {
      "from": "O2",
      "to": "E1",
    },
    Object {
      "from": "O2",
      "to": "E2",
    },
    Object {
      "from": "O2",
      "to": "O1",
    },
  ],
  "nodes": Array [
    Object {
      "id": "O1",
      "label": "O1",
    },
    Object {
      "id": "E1",
      "label": "E1",
    },
    Object {
      "id": "E2",
      "label": "E2",
    },
    Object {
      "id": "O2",
      "label": "O2",
    },
  ],
}
`;
